### PR TITLE
expose AuthEnc in the P2P config

### PR DESCRIPTION
**Pull request by melekes**
if AuthEnc is true, dialed peers must have a node ID in the address and
it must match the persistent pubkey from the secret handshake.

Refs #1157

### CODE REVIEW

**Comment by melekes** [comment]
```
2018-02-15T19:00:11Z box174.localdomain docker/local_testnet_4[14886]: E[02-15|19:00:11.510] Dialing failed                               module=p2p err="Failed to authentic
ate peer. Dialed 4c5f061dac28660853904a66705b12ca2b317572@172.57.0.103:46656, but got peer with ID e575418672e109ba8600715b02d45e359823907f"
```

**Comment by melekes** [comment]
Oh, that's because first ID was derived from pubkey, second - from `(SecretConnection).Pubkey` 

```
2018-02-15T19:00:11Z box174.localdomain docker/local_testnet_3[14886]: I[02-15|19:00:11.510] Peer handshake                               module=p2p peer=172.57.0.104:36607
peerNodeInfo="NodeInfo{pk: {<nil>}, moniker: , network:  [listen ], version:  ([])}"
```

but why `node_info` is empty???

**Comment by ebuchman** [comment]
Hmm, a peers NodeKey and the (SecretConnect).PubKey should be the same

**Comment by melekes** [comment]
I thought so too. Trying to understand why sometimes I get different IDs 

```
2018-02-19T08:00:47+00:00 172.17.0.1 local_testnet_1[1031]: E[02-19|08:00:47.677] Error dialing peer                           module=p2p err="Failed to authenticate peer. Dialed {99dbbd2afc28fb5bac5574afaf0d9c806ced3b55 172.57.0.102 46656 99dbbd2afc28fb5bac5574afaf0d9c806ced3b55@172.57.0.102:46656}, but got peer with ID ffe8642cfb46872355d5c07313a47800f0b06100"
```

**Comment by melekes** [comment]
```
PUBKEY FROM CONNECTION:
2018-02-19T09:21:42+00:00 172.17.0.1 local_testnet_1[1031]: fa981d707833b5aeb8bbcb8fd08073004b3776e2
2018-02-19T09:21:42+00:00 172.17.0.1 local_testnet_1[1031]: {"type":"ed25519","data":"B6C58FC656AF802053C453F6D85E067B9A2606794D8C2AAC1217AD605539F603"}

PREGENERATED PUBKEY (mach2):
99dbbd2afc28fb5bac5574afaf0d9c806ced3b55@172.57.0.102:46656,{"type":"ed25519","data":"AE67AC697D135AA0B4601EA57EAAB3FEBF4BAA4F229C45A598C2985B12FCD1A1"}
```

**Comment by melekes** [comment]
```
	// TODO: pass in like priv_val
	nodeKey, err := p2p.LoadOrGenNodeKey(n.config.NodeKeyFile())
	if err != nil {
		return err
	}
```

found it ^^^

File: CHANGELOG.md [hunk 1]
```diff
@@ -26,6 +26,12 @@ BUG FIXES:
 - Graceful handling/recovery for apps that have non-determinism or fail to halt
 - Graceful handling/recovery for violations of safety, or liveness
 
+## 0.16.1 (TBD)
+
+IMPROVEMENTS:
+- [config] exposed `auth_enc` flag to enable/disable encryption
+- [p2p] when `auth_enc` is true, all dialed peers must have a node ID (matching its pubkey)
```
Comment (melekes) [inline CHANGELOG.md]: "a node ID in their address"

File: cmd/tendermint/commands/show_node_id.go [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+var ShowNodeIDCmd = &cobra.Command{
+	Use:   "show_node_id",
+	Short: "Show this node's ID",
+	Run:   showNodeID,
+}
+
+func showNodeID(cmd *cobra.Command, args []string) {
+	nodeKey, err := p2p.LoadOrGenNodeKey(config.NodeKeyFile())
+	if err != nil {
+		logger.Error("Failed to load (or generate) node key", "err", err)
```
Comment (melekes) [inline cmd/tendermint/commands/show_node_id.go]: `"path", config.NodeKeyFile()`

File: docs/specification/configuration.rst [hunk 3]
```diff
@@ -121,6 +121,9 @@ like the file below, however, double check by inspecting the
     # Does not work if the peer-exchange reactor is disabled.
     seed_mode = false
 
+    # Authenticated encryption
+    auth_enc = {{ .P2P.AuthEnc }}
```
Comment (zramsay) [inline docs/specification/configuration.rst]: should be the default value, not template

**Comment by melekes** [comment]
Forgot to push node_id.json files.. Sorry

**Comment by melekes** [comment]
Hmmm... somewhere in our code we're saving addresses without an ID

```
2018-02-20T05:38:04Z box4450.localdomain docker/local_testnet_3[15473]: I[02-20|05:38:04.586] Added new address                            module=p2p book=/go/src/github.com
/tendermint/tendermint/test/p2p/data/mach3/core/config/addrbook.json address=f47c979740323445ce90a355b41d6f9f0d1b3347@172.57.0.103:46656 total=4
2018-02-20T05:38:04Z box4450.localdomain docker/local_testnet_3[15473]: I[02-20|05:38:04.586] Added new address                            module=p2p book=/go/src/github.com/tendermint/tendermint/test/p2p/data/mach3/core/config/addrbook.json address=172.57.0.102:46656 total=5
```

later if we try to connect, test fails because ID is now required

**Comment by ebuchman** [comment]
Closing for https://github.com/tendermint/tendermint/pull/1226

**Pull request closed without merging**